### PR TITLE
Migrate to async for .NET versions

### PR DESCRIPTION
**Pull request by nwithan8**
This has been a long time coming, but we're finally making our library asynchronous.

It's best practice in C# (and most programming languages, in fact) to make I/O bound tasks, such as network requests, [execute asynchronously](https://docs.microsoft.com/en-us/dotnet/csharp/async) so as to not lock up the main thread of an application. Our library previously executed HTTP requests synchronously, which needless blocked the application for several seconds, even the API call had already completed.

For example, in our own testing, we found that the `Shipment.Create()` function was taking almost 8 seconds to execute because the runtime was effectively padding the execution to give the HTTP call plenty of time to finish. When we migrated this function to asynchronous, the function returned the response from the HTTP call as soon as it was available, which was approximately 350 milliseconds. Another test, `Address.Create()` dropped from 2 seconds to ~400 milliseconds.

This PR:
- Migrates all functions in the .NET versions (Core 3.1, 5.0 and 6.0 currently) to asynchronous functions.
- <del>Our .NET Framework 3.5 version unfortunately cannot be migrated to asynchronous, as asynchronous functionality was not introduced until [.NET Framework 4.5](https://devblogs.microsoft.com/dotnet/async-in-4-5-worth-the-await/)</del>
- New .NET Standard 2.0 version is also asynchronous.
- Migrates all unit tests to asynchronous functions
- Fixed/enforce nullability of parameters in .NET

There's a joke that "async spreads like a virus", because when you make a function async/await, you have to make the function that called that function async/await, and the function that called that function... until all your functions are async or you reach a function that doesn't return anything.

Those implementing our now-asynchronous library may have varying levels of asynchronousity already in their applications, so this may be easier for some to implement than others. I recommend, for those unfamiliar, reading up on how asynchronous code works in C#, and how to [implement it into your final solution](https://exceptionnotfound.net/async-await-in-asp-net-csharp-ultimate-guide/).

If, for whatever reason, you must continue to run our API client library synchronously, you can use the built-in `.GetAwaiter().GetResult()` function to have any asynchronous function run synchronously. **This is [unadvised](https://www.nikouusitalo.com/blog/why-is-getawaiter-getresult-bad-in-c/), and will result in severe performance degradation.**

### CODE REVIEW

**Comment by Justintime50** [comment]
Will review this once we adjust the build matrix for Standard.

File: EasyPost/Client.cs [hunk 1]
```diff
@@ -51,7 +52,7 @@ public Client(ClientConfiguration clientConfiguration, HttpClient? customHttpCli
             {
                 Assembly assembly = typeof(Client).Assembly;
                 FileVersionInfo info = FileVersionInfo.GetVersionInfo(assembly.Location);
-                _libraryVersion = info.FileVersion;
+                _libraryVersion = info.FileVersion ?? "Unknown";
```
Comment (Justintime50) [inline EasyPost/Client.cs:55]: How did this sneak in or was it added? Rebase gone wrong somewhere? I thought this was already in the repo.

File: EasyPost/IResource.cs [hunk 2]
```diff
@@ -4,7 +4,7 @@ namespace EasyPost
 {
     public interface IResource
     {
-        Dictionary<string, object> AsDictionary();
+        Dictionary<string, object?> AsDictionary();
```
Comment (Justintime50) [inline EasyPost/IResource.cs:7]: What's with the proliferation of all these question marks throughout this PR? I thought this was only to make things async?

File: EasyPost/Insurance.cs [hunk 3]
```diff
@@ -98,33 +99,20 @@ public static Insurance Create(Dictionary<string, object> parameters)
...
-        public static Insurance Retrieve(string id)
+        public static async Task<Insurance> Retrieve(string id)
         {
             Request request = new Request("insurances/{id}");
             request.AddUrlSegment("id", id);
 
-            return request.Execute<Insurance>();
-        }
-
-        private static Insurance SendCreate(Dictionary<string, object> parameters)
```
Comment (Justintime50) [inline EasyPost/Insurance.cs:117]: Please call the removal of this out in the CHANGELOG.

File: EasyPost/Order.cs [hunk 4]
```diff
@@ -127,25 +128,12 @@ public static Order Create(Dictionary<string, object> parameters)
...
-        public static Order Retrieve(string id)
+        public static async Task<Order> Retrieve(string id)
         {
             Request request = new Request("orders/{id}");
             request.AddUrlSegment("id", id);
 
-            return request.Execute<Order>();
-        }
-
-        private static Order SendCreate(Dictionary<string, object> parameters)
```
Comment (Justintime50) [inline EasyPost/Order.cs:138]: Please call the removal of this out in the CHANGELOG.

File: EasyPost/Tracker.cs [hunk 5]
```diff
@@ -102,31 +103,30 @@ public static TrackerCollection All(Dictionary<string, object> parameters = null
...
             request.AddBody(new Dictionary<string, object>
             {
                 {
                     "trackers", parameters
                 }
             });
-            request.Execute<Tracker>();
+            return await request.Execute();
             // This endpoint does not return a response so we return true here
-            return true;
```
Comment (Justintime50) [inline EasyPost/Tracker.cs:129]: Why are we switching this?

File: EasyPost/Insurance.cs (hunk 3 above)
Comment (nwithan8) [reply EasyPost/Insurance.cs:117]: It was internal, so no one ever saw it. Dead code.

File: EasyPost/Order.cs (hunk 4 above)
Comment (nwithan8) [reply EasyPost/Order.cs:138]: It was internal, so no one ever saw it. Dead code.

File: EasyPost/Tracker.cs (hunk 5 above)
Comment (nwithan8) [reply EasyPost/Tracker.cs:129]: Instead of always getting `true` back from the function, we now actually get a boolean based on whether the API call was successful or not.

File: EasyPost/IResource.cs (hunk 2 above)
Comment (nwithan8) [reply EasyPost/IResource.cs:7]: The `?`  in C# means nullable. The IDE is constantly complaining that we are setting `null` as a default value to non-nullable values. So now, the values are nullable. More explicit to the end user, less likely to throw a runtime error.

File: EasyPost/Client.cs (hunk 1 above)
Comment (nwithan8) [reply EasyPost/Client.cs:55]: I believe I added that in earlier, but it's just a fallback in case the .FileVersion function fails (returns null, but `_libraryVersion` can't have a null value assigned to it)

**Comment by jmalatia** [comment]
Being that you guys are finally giving this C# library some much needed attention, you really should address:

- [Follow C# Naming styles](https://github.com/EasyPost/easypost-csharp/issues/66#issuecomment-270287354)... use PascalCase for naming classes and public properties
- Get rid of dictionaries on the public side of the API
- DI friendly

Many of the same things echoed for years... see this post https://github.com/EasyPost/easypost-csharp/issues/66#issuecomment-270287354 from 2017

see this library: https://github.com/kendallb/easypost-async-csharp

File: EasyPost/Tracker.cs [hunk 6]
```diff
@@ -102,31 +103,30 @@ public static TrackerCollection All(Dictionary<string, object> parameters = null
...
-            Request request = new Request("trackers/create_list", RestSharp.Method.Post);
+            Request request = new Request("trackers/create_list", Method.Post);
             request.AddBody(new Dictionary<string, object>
             {
                 {
                     "trackers", parameters
                 }
             });
-            request.Execute<Tracker>();
+            return await request.Execute();
```
Comment (jchen293) [inline EasyPost/Tracker.cs:128]: Quick question, should this be like `return await request.Execute<Tracker>();`?

File: EasyPost.Tests/ShipmentTest.cs [hunk 7]
```diff
@@ -14,11 +15,11 @@ public void Initialize()
         }
 
         [TestMethod]
-        public void TestCreate()
+        public async Task TestCreate()
         {
-            VCR.Replay("create");
+            VCR.Record("create");
```
Comment (jchen293) [inline EasyPost.Tests/ShipmentTest.cs:20]: Why was this changed from `Replay` to `Record. I don't have much knowledge on scotch VCR yet, but this change seems inconsistent with other unit tests.

**Review by jchen293 — COMMENTED** [review]
Code looks good, have you re-record cassettes locally to see if all unit tests pass? @nwithan8

File: EasyPost/Tracker.cs (hunk 6 above)
Comment (nwithan8) [reply EasyPost/Tracker.cs:128]: No, this API endpoint doesn't return any data. `request.Execute()` returns true/false based on whether the API returned an error status code, while `request.Execute<T>()` returns a T-type object.

File: EasyPost.Tests/ShipmentTest.cs (hunk 7 above)
Comment (nwithan8) [reply EasyPost.Tests/ShipmentTest.cs]: Resolved in [13dec76](https://github.com/EasyPost/easypost-csharp/pull/235/commits/13dec7625f11be9e9d01d625178bc7e0fd723410)

**Comment by nwithan8** [comment]
> Code looks good, have you re-record cassettes locally to see if all unit tests pass? @nwithan8

I just deleted and re-recorded all cassettes locally and "everything" works (two batch tests failed because of the timing issue, as expected. Since those cassettes couldn't be finished, I didn't push new cassettes to the repo)

**Review by jchen293 — APPROVED** [review]
nice work
